specs/DataSet.rst

```diff
@@ -0,0 +1,253 @@
+   It should be possible to read "new data" in a DataSet; that is, to read everything after a cursor.
+#. It should be possible to subscribe to change notifications from a DataSet.
+   It is acceptable if such subscriptions must be in-process until QCoDeS multiprocessing is redone.
+   Change notifications should include the results that were added to the DataSet that triggered the notification.
+
+Storage and Persistence
+-----------------------
+
+#. A DataSet object should allow writing to and reading from storage in a variety of formats.
+#. Users should be able to define new persistence formats.
+#. Users should be able to specify where a DataSet is written.
+
+Interface
+=========
+
+Creation
+--------
+
+ParamSpec
+~~~~~~~~~
+
+A ParamSpec object specifies a single parameter in a DataSet.
+
+ParamSpec(Parameter p, optional=)
+    Creates a parameter specification from a QCoDeS Parameter.
+    If optional is provided and is true, then the parameter is optional in each result.
+    
+ParamSpec(name, role, type, desc=, optional=)
+    Creates a parameter specification with the given name, role (‘I’ or ‘O’), and type. 
+    The type should be a NumPy dtype object.
+    If a description is provided, it is included in the metadata of the DataSet.
+    The description can be a simple string or a string-to-string dictionary.
+    If optional is provided and is true, then the parameter is optional in each result.
+
+DataSet
+~~~~~~~
+
+DataSet()
+    Creates a DataSet with no parameters.
+
+DataSet(specs)
+    Creates a DataSet for the provided list of parameter specifications.
+    Each item in the list should either be a QCoDeS Parameter, a tuple of a Parameter and a Boolean, or a ParamSpec object.
+    A Parameter or a Parameter tupled with a false value indicates a required parameter; a Parameter tupled with a true value indicates an optional parameter.
+
+DataSet.add_parameter(spec)
+    Adds a parameter to an existing DataSet.
+    The spec should either be a QCoDeS Parameter, a tuple of a Parameter and a Boolean, or a ParamSpec object.
+    A Parameter or a Parameter tupled with a false value indicates a required parameter; a Parameter tupled with a true value indicates an optional parameter.
+    It is an error to add a parameter to a non-empty DataSet.
+
+DataSet.add_parameters(specs)
+    Adds a list of parameters to an existing DataSet.
+    Each item in the list should either be a QCoDeS Parameter, a tuple of a Parameter and a Boolean, or a ParamSpec object.
+    A Parameter or a Parameter tupled with a false value indicates a required parameter; a Parameter tupled with a true value indicates an optional parameter.
+    It is an error to add a parameter to a non-empty DataSet.
+
+DataSet.add_metadata(tag=, info=)
+    Adds metadata to the current DataSet.
+    The metadata is stored under the provided tag.
+    It is an error to add metadata to a completed DataSet.
+
+Writing
+-------
+
+DataSet.add_result(**kwargs)
```

Comment:
I think all of the keywords would be parameter names, so not known at compile time anyway. At least, that was what I had in mind. That means that each DataSet might have different arguments to add_result.

The alternative would be for add_result to take a single (positional) parameter that is a dictionary of parameter name/parameter value pairs. I'm fine with that, but it's no less opaque than just keyword arguments.

If there's a better way, please tell me!! My last heavy Python coding was in version 1.4...